react-components/src/query/useSearchMappedEquipmentAssetMappings.tsx

```diff
@@ -48,51 +49,69 @@ export const useSearchMappedEquipmentAssetMappings = (
-    queryKey: [
+  const assetIds = useMemo(() => {
+    if (assetMappingList === undefined) return new Set<number>();
+    return new Set(
+      assetMappingList.flatMap((mapping) =>
+        mapping.assetMappings.map((assetMapping) => assetMapping.assetId)
+      )
+    );
+  }, [assetMappingList]);
+
+  const [unmappedAssetIds, setUnmappedAssetIds] = useState(new Set<number>());
+
+  const queryKey = useMemo(
+    () => [
       'reveal',
       'react-components',
       'search-mapped-asset-mappings',
       query,
-      limit,
       ...models.map((model) => [model.modelId, model.revisionId])
     ],
-    queryFn: async ({ pageParam }: { pageParam: string | undefined }) => {
-      if (query === '' || assetMappingList === undefined) {
-        return { assets: [], nextCursor: undefined };
-      }
+    [query, models]
+  );
+
+  const fetchAssets = async (
+    cursor: string | undefined,
+    accumulatedAssets: Asset[]
+  ): Promise<{ assets: Asset[]; nextCursor: string | undefined }> => {
+    const assetsResponse = await getAssetsList(sdk, {
+      query,
+      limit: 1000,
+      cursor
+    });
 
-      const fetchAssets = async (
-        cursor: string | undefined,
-        accumulatedAssets: Asset[]
-      ): Promise<{ assets: Asset[]; nextCursor: string | undefined }> => {
-        const assetsResponse = await getAssetsList(sdk, {
-          query,
-          limit,
-          cursor
-        });
+    const fetchedAssets = assetsResponse.items.filter(isDefined);
+    const filteredSearchedAssets = fetchedAssets.filter(
+      (asset) => assetIds.has(asset.id) && !unmappedAssetIds.has(asset.id)
+    );
 
-        const fetchedAssets = assetsResponse.items.filter(isDefined);
-        const filteredSearchedAssets = assetMappingList.flatMap((mapping) => {
-          return mapping.assetMappings
-            .filter((assetMapping) =>
-              fetchedAssets.some((asset) => asset.id === assetMapping.assetId)
-            )
-            .map((assetMapping) => fetchedAssets.find((asset) => asset.id === assetMapping.assetId))
-            .filter(isDefined);
-        });
+    const uniqueAssets = uniqBy(
+      [...accumulatedAssets, ...filteredSearchedAssets],
+      (asset) => asset.id
+    );
 
-        const uniqueAssets = uniqBy(
-          [...accumulatedAssets, ...filteredSearchedAssets],
-          (asset) => asset.id
-        );
+    if (uniqueAssets.length >= limit || assetsResponse.nextCursor === undefined) {
+      return { assets: uniqueAssets, nextCursor: assetsResponse.nextCursor };
+    }
 
-        if (uniqueAssets.length >= limit || assetsResponse.nextCursor === undefined) {
-          return { assets: uniqueAssets, nextCursor: assetsResponse.nextCursor };
-        }
+    const newUnmappedAssetIds = new Set(unmappedAssetIds);
+    fetchedAssets.forEach((asset) => {
+      if (!assetIds.has(asset.id)) {
+        newUnmappedAssetIds.add(asset.id);
+      }
+    });
+    setUnmappedAssetIds(newUnmappedAssetIds);
```

Comment:
Do we need to create a new Set in every iteration? We could just keep the same Set and add into it? 
That would also mean we don't need a `useState`, but could use a `useRef` instead. Unless we do the `useCallback` suggestion above, then it can just be a local variable